### DISCUSSION TITLE
Constexpr SYCL vectors

### DISCUSSION THREAD

**Discussion by Michoumichmich** [AdaptiveCpp/AdaptiveCpp#678] [Q&A] [answered]
Hi, 
Are there some technical reasons why the `sycl::vec<>` constructors are not marked `constexpr` in hipSYCL? In the specification they are constexpr. That would be especially useful to initialise global data. Right now a const global array containing  `sycl::vec<>` won't be initialised at compile time in hipSYCL and so won't be available in the GPU (on the CUDA backend at least).
Best, 
Michel

**Comment by illuhad** [comment, marked as answer]
They were not `constexpr` in earlier revisions of the spec ;) I believe @DieGoldeneEnte has done some work on this - if I recall correctly, there are some additional changes required, so it might not be as easy as just adding `constexpr` in front of them.

**Comment by DieGoldeneEnte** [comment]
The changes required to make SYCL vectos `constexpr` are in #628 (needs to be rebased and is build on top of #578, so can't be merged directly).
It mostly comes down to adding static casts to some buildins macros and some adjustments in the constructor since member functions can't be called before the class is initialized (this concerns `vec::count_num_elements`) or something along those lines.

**Reply by Michoumichmich** [reply]
Awesome! About `count_num_elements`, shouldn't that be a `else if constexpr` here ? (saw that while looking into it before): https://github.com/illuhad/hipSYCL/blob/ee8cd86f95fe5604b6cf130d724828f7f5f0e121/include/hipSYCL/sycl/libkernel/vec.hpp#L853

**Reply by DieGoldeneEnte** [reply]
Yes, good catch! I assumed it wasn't necessary, but it is. Thanks

**Reply by Michoumichmich** [reply]
> Yes, good catch! I assumed it wasn't necessary, but it is. Thanks

Oh no, my bad, turns out neither of them is necessary as it don't change the "constexprness" of the function^^
